libs/smart-contracts/src/contracts/token.ts

```diff
@@ -0,0 +1,45 @@
+import type { BigNumber } from 'ethers';
+import { ethers } from 'ethers';
+import erc20Abi from '../abis/erc20_abi.json';
+import erc20AbiFaucet from '../abis/erc20_abi_faucet.json';
+
+export class Token {
+  public contract: ethers.Contract;
+  public faucetable: boolean;
+
+  constructor(
+    address: string,
+    signerOrProvider: ethers.Signer | ethers.providers.Provider,
+    faucetable = false
+  ) {
+    this.contract = new ethers.Contract(
+      address,
+      faucetable ? erc20AbiFaucet : erc20Abi,
+      signerOrProvider
+    );
+    this.faucetable = faucetable;
+  }
+
+  totalSupply() {
+    return this.contract.totalSupply();
+  }
+  balanceOf(account: string): Promise<BigNumber> {
+    return this.contract.balanceOf(account);
+  }
+  allowance(owner: string, spender: string): Promise<BigNumber> {
+    return this.contract.allowance(owner, spender);
+  }
+  approve(spender: string, amount: string) {
+    return this.contract.approve(spender, amount);
+  }
+  decimals(): Promise<number> {
+    return this.contract.decimals();
+  }
+  faucet() {
+    if (!this.faucetable) {
+      throw new Error('Cannot use faucet');
+    }
+
+    return this.contract.faucet();
+  }
```

Comment:
nitpick (non-blocking): Not sure about having fauctable as a prop, could we use one of:
1. Detect if the method is present?
2. Let the eth transaction fail and catch the error?
3. Have a separate class as these are two different contracts and so should have differing thin wrappers?

Reply:
I think we have to go with option 3 here. The ethers lib will only add the method to the instance of ethers.Contract if it exists in the ABI, which means we have to pass in the ABI to the constructor which I don't think is desirable. Having an additional wrapper makes most sense given with our future plans here.